mlir/include/mlir/Dialect/Utils/IndexingUtils.h

```diff
@@ -202,6 +202,9 @@ SmallVector<T> applyPermutation(ArrayRef<T> input,
                                 ArrayRef<int64_t> permutation) {
   assert(input.size() == permutation.size() &&
          "expected input rank to equal permutation rank");
+  assert(
```

Comment:
Would you be able to change the result of this to be a `FailureOr<SmallVector<T>>` That way, we can avoid the asserts

Reply:
That change can be done. But that would lead to :
1) Changing the code of all the callers of "applyPermutation()" to take care of the failue path. 
2) Each caller can throw different error message on failure, which can lead to bad user experience. 

The original aim of this patch was to throw a meaningful error message. 
I can change the function signature in a new patch if all the reviewers agrees for that.